zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/S3NotebookRepo.java

```diff
@@ -70,69 +74,95 @@
   //  3. Credential profiles file at the default location (~/.aws/credentials)
   //       shared by all AWS SDKs and the AWS CLI
   //  4. Instance profile credentials delivered through the Amazon EC2 metadata service
-  private AmazonS3 s3client = new AmazonS3Client(new DefaultAWSCredentialsProviderChain());
-  private static String bucketName = "";
-  private static String endpoint = "";
-  private String user = "";
-
-  private ZeppelinConfiguration conf;
+  private final AmazonS3 s3client;
+  private final String bucketName;
+  private final String user;
+  private final ZeppelinConfiguration conf;
 
   public S3NotebookRepo(ZeppelinConfiguration conf) throws IOException {
     this.conf = conf;
     bucketName = conf.getBucketName();
-    endpoint = conf.getEndpoint();
     user = conf.getUser();
-    
-    s3client.setEndpoint(endpoint);
+
+    // always use the default provider chain
+    AWSCredentialsProvider credentialsProvider = new DefaultAWSCredentialsProviderChain();
+
+    // see if we should be encrypting data in S3
+    String kmsKeyID = conf.getS3KMSKeyID();
+    if (kmsKeyID != null) {
+      // use the AWS KMS to encrypt data
+      KMSEncryptionMaterialsProvider emp = new KMSEncryptionMaterialsProvider(kmsKeyID);
+      this.s3client = new AmazonS3EncryptionClient(credentialsProvider, emp);
+    }
+    else if (conf.getS3EncryptionMaterialsProviderClass() != null) {
+      // use a custom encryption materials provider class
+      EncryptionMaterialsProvider emp = createCustomProvider(conf);
+      this.s3client = new AmazonS3EncryptionClient(credentialsProvider, emp);
+    }
+    else {
+      // regular S3
+      this.s3client = new AmazonS3Client(credentialsProvider);
+    }
+
+    // set S3 endpoint to use
+    s3client.setEndpoint(conf.getEndpoint());
+  }
+
+  /**
+   * Create an instance of a custom encryption materials provider class
+   * which supplies encryption keys to use when reading/writing data in S3.
+   */
+  private EncryptionMaterialsProvider createCustomProvider(ZeppelinConfiguration conf) {
+    // use a custom encryption materials provider class
+    String empClassname = conf.getS3EncryptionMaterialsProviderClass();
+    EncryptionMaterialsProvider emp;
+    try {
+      Object empInstance = Class.forName(empClassname).newInstance();
+      if (empInstance instanceof EncryptionMaterialsProvider) {
+        emp = (EncryptionMaterialsProvider) empInstance;
+      }
+      else {
+        throw new IllegalArgumentException("Class " + empClassname + " does not implement "
```

Comment:
Why not to throw `IOException` since constructor is already throwing it and upper layer may rely on that